### PR TITLE
fix(pointcloud_preprocessor): changed pointcloud conversion method

### PR DESCRIPTION
**Pull request by 0x126**
Signed-off-by: Shinnosuke Hirakawa <shinnosuke.hirakawa@tier4.jp>

### #Background
 - For concatenation, all pointclouds must be converted to Pointcloud2 with the same point_type
 - The concatenated pointcloud is  used not only for localization and planning, but also as input to the perception modules, so in addition to `XYZ`, `intensity` is required
 - Currently, the conversion is done using PCL functions, as shown in the code below.
```
pcl::PointCloud<PointXYZI> tmp_cloud;
pcl::fromROSMsg(input_cloud, tmp_cloud);
pcl::toROSMsg(tmp_cloud, output_cloud);
output_cloud.header = input_cloud.header;
```
 - The above PCL funtions works, but if the input pointcloud does not contain an intensity, it will output the error shown below
` Failed to find match for field 'intensity'.`

### #What this PR changes
 - Convert pointclouds without using PCL functions.
 - Check for the presence of "intensity" and branch the condition.
 
## Review Procedure(required)
 - Check the concatenated pointcloud output
 - Input the concatenated pointcloud into the perception modules and check the output
 - Do the above for pointclouds with/without intensity.

## Related PR(optional)

https://github.com/autowarefoundation/autoware.universe/pull/308
## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: sensing/pointcloud_preprocessor/src/concatenate_data/concatenate_data_nodelet.cpp [hunk 1]
```diff
@@ -323,22 +351,20 @@ void PointCloudConcatenateDataSynchronizerComponent::setPeriod(const int64_t new
 }
 
 void PointCloudConcatenateDataSynchronizerComponent::cloud_callback(
-  const sensor_msgs::msg::PointCloud2::ConstSharedPtr & input_ptr, const std::string & topic_name)
+  const sensor_msgs::msg::PointCloud2::SharedPtr & input_ptr, const std::string & topic_name)
```
Comment (satoshi-ota) [inline sensing/pointcloud_preprocessor/src/concatenate_data/concatenate_data_nodelet.cpp:354]: I think this diff is **NOT** necessary :thinking: 

FYI
https://github.com
Screenshot from 2022-02-03 09-53-16
/tier4/AutowareArchitectureProposal.iv/pull/792/files

Comment (satoshi-ota) [reply sensing/pointcloud_preprocessor/src/concatenate_data/concatenate_data_nodelet.cpp:354]: You're not using `reinterpret_cast`, are you?

**Comment by satoshi-ota** [comment]
We have checked the operation on the actual vehicle. 

And, it seems to be working correctly on logging_simulator by using rosbag (2022-02-02-1544_ps1_300kg_auto_left_dual_n4_2). :+1: 

```sh
ros2 launch autoware_launch logging_simulator.launch.xml map_path:=/home/satoshi/map/sagamihara_glp vehicle_model:=gsm8 sensor_model:=aip_x2 planning:=false control:=false localization:=false
```

- `Failed to find match for field 'intensity'.` don't shown up.
- concatenate_filter and dual_retrun_outlier_filter is work correctly.
- perception also work correctly.

rviz_screenshot_2022_02_03-10_52_41

File: sensing/pointcloud_preprocessor/src/concatenate_data/concatenate_data_nodelet.cpp (hunk 1 above)
Comment (0x126) [reply sensing/pointcloud_preprocessor/src/concatenate_data/concatenate_data_nodelet.cpp:354]: I can not use `PointCloud2Iterator` with `sensor_msgs::msg::PointCloud2::ConstSharedPtr`.
Do you have any solution?

Comment (satoshi-ota) [reply sensing/pointcloud_preprocessor/src/concatenate_data/concatenate_data_nodelet.cpp:354]: OK, I see. I understand that It is necessary for `PointCloud2Iterator` :+1:

**Review by satoshi-ota — APPROVED** [review]
@0x126 It looks good to me. Please write release note :pray:

**Comment by 0x126** [comment]
@1222-takeshi Can you try this PR with AIP X1 settings？

**Review by 1222-takeshi — APPROVED** [review]
LGTM
I checked it in the x1 settings and it seemed to concatenate fine.

**Pull request merged**